File: apps/social/before/lib/logger.go
package lib

import (
	"encoding/json"
	"fmt"
	"time"
)

type LogEntry struct {
	Timestamp string                 `json:"timestamp"`
	Level     string                 `json:"level"`
	Event     string                 `json:"event"`
	TraceID   string                 `json:"traceId"`
	SpanID    string                 `json:"spanId"`
	Fields    map[string]interface{} `json:"fields"`
}

var globalTraceID = generateUUID()

func generateUUID() string {
	// Simple UUID v4-like generator
	return fmt.Sprintf("%08x-%04x-%04x-%04x-%012x",
		randInt(), randInt()%65536, randInt()%65536, randInt()%65536, randInt())
}

func randInt() int {
	return int(time.Now().UnixNano() / 1e6 % 1e8)
}

type logger struct{}

func (l *logger) log(level, event string, fields map[string]interface{}) {
	entry := LogEntry{
		Timestamp: time.Now().UTC().Format(time.RFC3339Nano),
		Level:     level,
		Event:     event,
		TraceID:   globalTraceID,
		SpanID:    generateUUID(),
		Fields:    fields,
	}
	data, _ := json.Marshal(entry)
	fmt.Println(string(data))
}

func (l *logger) Info(event string, fields map[string]interface{}) {
	l.log("info", event, fields)
}

func (l *logger) Warn(event string, fields map[string]interface{}) {
	l.log("warn", event, fields)
}

func (l *logger) Error(event string, fields map[string]interface{}) {
	l.log("error", event, fields)
}

func (l *logger) Debug(event string, fields map[string]interface{}) {
	l.log("debug", event, fields)
}

var Logger = &logger{}

File: apps/fintech/after/lib/logger.py
"""Structured business logger.

Every log line goes two places:
  1. stdout as a single JSON line (unchanged behavior), and
  2. the stdlib logger "hello-otel-fintech", which otel_init attaches an
     OpenTelemetry LoggingHandler to — so records export over OTLP gRPC
     with trace correlation (trace_id/span_id from the active span, which
     FastAPIInstrumentor provides inside request handlers).
"""

import json
import logging
from datetime import datetime

from opentelemetry import trace as _otel_trace

from lib.events import record_event

_LEVELS = {
    "debug": logging.DEBUG,
    "info": logging.INFO,
    "warn": logging.WARNING,
    "error": logging.ERROR,
}

# OTLP handler is attached to this logger by otel_init.init_otel().
_py_logger = logging.getLogger("hello-otel-fintech")
_py_logger.setLevel(logging.DEBUG)
_py_logger.propagate = False


def _current_trace_ids():
    ctx = _otel_trace.get_current_span().get_span_context()
    if ctx.is_valid:
        return format(ctx.trace_id, "032x"), format(ctx.span_id, "016x")
    return "", ""


def log(level, event, fields=None):
    if fields is None:
        fields = {}

    trace_id, span_id = _current_trace_ids()
    entry = {
        "timestamp": datetime.utcnow().isoformat() + "Z",
        "level": level,
        "event": event,
        "traceId": trace_id,
        "spanId": span_id,
        "fields": fields,
    }
    print(json.dumps(entry))
    record_event(level, event, fields)

    # Ship over OTLP with the business fields as log attributes. The
    # LoggingHandler adds trace context from the active span automatically.
    attrs = {
        f"app.{k}": (v if isinstance(v, (str, int, float, bool)) else str(v))
        for k, v in fields.items()
        if v is not None
    }
    _py_logger.log(_LEVELS.get(level, logging.INFO), event, extra=attrs)


class Logger:
    def debug(self, event, fields=None):
        log("debug", event, fields)

    def info(self, event, fields=None):
        log("info", event, fields)

    def warn(self, event, fields=None):
        log("warn", event, fields)

    def error(self, event, fields=None):
        log("error", event, fields)


logger = Logger()
